### PR TITLE
Adding certificate validity duration in Kafka user

### PR DESCRIPTION
**Pull request by shubhamcoc**
## Description

This MR is to add duration in Kafka user spec. It allows user to specify validity time for certificate created by cert manager.  
It helps the user to create user certificate for longer duration than the default value of cert manager. #964 

## Type of Change

- [x] New Feature

## Checklist

- [x] I have read the [contributing guidelines](../CONTRIBUTING.md)
- [x] Existing issues have been referenced (where applicable)
- [x] All code style checks pass
- [x] New code contribution is covered by automated tests
- [x] All new and existing tests pass

### CODE REVIEW

**Review by bartam1 — CHANGES_REQUESTED** [review]
We also have to use the validation duration for the k8s CSR object:
https://github.com/banzaicloud/koperator/blob/a64633a24935763cab1714e909e669323392ffe8/pkg/pki/k8scsrpki/k8scsr_user.go#L253

**Comment by shubhamcoc** [comment]
> We also have to use the validation duration for the k8s CSR object:
> 
> https://github.com/banzaicloud/koperator/blob/a64633a24935763cab1714e909e669323392ffe8/pkg/pki/k8scsrpki/k8scsr_user.go#L253

Yes one validation is required as we have 2 different combination available. One for cert-manager with 1 hour as minimum duration and another with k8s CSR object with 10 mins as minimum duration.

**Comment by shubhamcoc** [comment]
> We also have to use the validation duration for the k8s CSR object:
> 
> https://github.com/banzaicloud/koperator/blob/a64633a24935763cab1714e909e669323392ffe8/pkg/pki/k8scsrpki/k8scsr_user.go#L253

**Review by panyuenlau — COMMENTED** [review]
Can we also extend the existing tests to include the change in this PR?

https://github.com/banzaicloud/koperator/blob/ffb1fabbf3c44e6069b6dc51d13e55a82f11f986/pkg/pki/certmanagerpki/certmanager_user_test.go#L66

https://github.com/banzaicloud/koperator/blob/ffb1fabbf3c44e6069b6dc51d13e55a82f11f986/pkg/pki/k8scsrpki/k8scsr_user_test.go#L91

File: controllers/cruisecontroloperation_controller.go [hunk 1]
```diff
@@ -267,6 +267,8 @@ func (r *CruiseControlOperationReconciler) executeOperation(ctx context.Context,
 		cruseControlTaskResult, err = r.scaler.RebalanceWithParams(ctx, ccOperationExecution.CurrentTaskParameters())
 	case banzaiv1alpha1.OperationStopExecution:
 		cruseControlTaskResult, err = r.scaler.StopExecution(ctx)
+	case banzaiv1alpha1.OperationStatus:
+		err = errors.NewWithDetails("Cruise Control operation not supported", "name", ccOperationExecution.GetName(), "namespace", ccOperationExecution.GetNamespace(), "operation", ccOperationExecution.CurrentTaskOperation(), "parameters", ccOperationExecution.CurrentTaskParameters())
```
Comment (panyuenlau) [inline controllers/cruisecontroloperation_controller.go:271]: Note: these two lines are not related to the PR and they are copied from https://github.com/banzaicloud/koperator/pull/974 to make linter not complain the missing case in the switch statement.

**Comment by shubhamcoc** [comment]
> Can we also extend the existing tests to include the change in this PR?
> 
> https://github.com/banzaicloud/koperator/blob/ffb1fabbf3c44e6069b6dc51d13e55a82f11f986/pkg/pki/certmanagerpki/certmanager_user_test.go#L66
> 
> https://github.com/banzaicloud/koperator/blob/ffb1fabbf3c44e6069b6dc51d13e55a82f11f986/pkg/pki/k8scsrpki/k8scsr_user_test.go#L91

Added it. Kindly review it

File: pkg/pki/certmanagerpki/certmanager_user_test.go [hunk 2]
```diff
@@ -103,6 +104,14 @@ func TestReconcileUserCertificate(t *testing.T) {
 		t.Error("Expected no error, got:", err)
 	}
 
+	// Test cert duration
+	user = newMockUser()
+	user.Spec.ExpirationSeconds = util.Int32Pointer(7200)
+
+	if _, err := manager.ReconcileUserCertificate(ctx, user, scheme.Scheme, clusterDomain); err != nil {
+		t.Error("Expected no error, got:", err)
+	}
```
Comment (panyuenlau) [inline pkg/pki/certmanagerpki/certmanager_user_test.go:113]: Instead of only verifying err, can we also verify the actual cert duration (IMO this is the most important part to verify here)?

Something like this as a reference:
```suggestion
	userCert, err := manager.ReconcileUserCertificate(ctx, user, scheme.Scheme, clusterDomain)
	if err != nil {
		t.Error("Expected no error, got:", err)
	}
	
	decodedCert, err := certutil.DecodeCertificate(userCert.Certificate)
	if err != nil {
		t.Error("Expected no error, got:", err)
	}
	
	if user.Spec.ExpirationSeconds != cert.NotAfter.Sub(cert.NotBefore).Seconds() {
	     t.Error("Expected cert duration is not the same as the actual cert duration, expected: %d, got: %d", user.Spec.ExpirationSeconds, cert.NotAfter.Sub(cert.NotBefore).Seconds())
	}
```

Comment (shubhamcoc) [reply pkg/pki/certmanagerpki/certmanager_user_test.go]: Hi @panyuenlau, it seems if we compare the values, it is not working because time related fields are not populated properly in test env, NotAfter and NotBefore both have zero value of time.Time (i.e. 0001-01-01 00:00:00 +0000 UTC) and the difference is coming as zero. I think we have to check for no err for now. What do you think?

Comment (shubhamcoc) [reply pkg/pki/certmanagerpki/certmanager_user_test.go:113]: Reverted the change.

**Review by panyuenlau — APPROVED** [review]

**Review by bartam1 — APPROVED** [review]
